### PR TITLE
feat: edfSchduler can be faster by maintaining heap manually (#1919)

### PR DESCRIPTION
**Pull request by jizhuozhi**
### Issues associated with this PR

See #1919 

### Solutions
I provide two heap implementations separately.

One of them is more general, named fastheap, and the parent and child elements are exchanged through the interface{} comparison function and single assignment.

And the other is to directly provide type-specific heap for edfEntry on the basis of single assignment exchange of parent and child elements

### Benchmark
```
// heap.Fix (Old)
goos: darwin
goarch: amd64
pkg: mosn.io/mosn/pkg/upstream/cluster
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_edfSchduler_NextAndPush
Benchmark_edfSchduler_NextAndPush/10-hosts
Benchmark_edfSchduler_NextAndPush/10-hosts-12         	 6653929	       164.6 ns/op
Benchmark_edfSchduler_NextAndPush/100-hosts
Benchmark_edfSchduler_NextAndPush/100-hosts-12        	 5170723	       248.1 ns/op
Benchmark_edfSchduler_NextAndPush/1000-hosts
Benchmark_edfSchduler_NextAndPush/1000-hosts-12       	 3475216	       316.0 ns/op
PASS

// type-specific edfHeap (Current)
goos: darwin
goarch: amd64
pkg: mosn.io/mosn/pkg/upstream/cluster
cpu: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Benchmark_edfSchduler_NextAndPush
Benchmark_edfSchduler_NextAndPush/10-hosts
Benchmark_edfSchduler_NextAndPush/10-hosts-12         	 9978801	       110.6 ns/op
Benchmark_edfSchduler_NextAndPush/100-hosts
Benchmark_edfSchduler_NextAndPush/100-hosts-12        	 8934966	       135.7 ns/op
Benchmark_edfSchduler_NextAndPush/1000-hosts
Benchmark_edfSchduler_NextAndPush/1000-hosts-12       	 6224355	       172.0 ns/op
PASS
```

### CODE REVIEW

File: pkg/upstream/cluster/edfheap.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+			newCap = minCap
+		}
+		oldElements := h.elements
+		h.elements = make([]*edfEntry, newCap)
+		copy(h.elements, oldElements)
+	}
+}
+
+// assertHeapOrdering for testing
+func (h *edfHeap) assertHeapOrdering(onFail func(i int, element interface{})) {
```
Comment (alpha-baby) [inline pkg/upstream/cluster/edfheap.go:151]: move to *_test.go file

File: pkg/upstream/cluster/edfheap.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+func (h *edfHeap) ensureIncrement() {
+	if h.size+1 > cap(h.elements) {
+		oldElements := h.elements
+		h.elements = make([]*edfEntry, 2*cap(h.elements))
+		copy(h.elements, oldElements)
+	}
+}
+
+func (h *edfHeap) ensureDecrement() {
+	if h.size*2 < cap(h.elements) {
```
Comment (alpha-baby) [inline pkg/upstream/cluster/edfheap.go:139]: should:

```go
if minCap < h.size && h.size*2 < cap(h.elements) {
//.....
}
```

File: pkg/upstream/cluster/edfheap.go [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+func (h *edfHeap) Size() int {
+	return h.size
+}
+
+func (h *edfHeap) Empty() bool {
+	return h.size == 0
+}
+
+func (h *edfHeap) fixUp(i int) bool {
```
Comment (alpha-baby) [inline pkg/upstream/cluster/edfheap.go:85]: need code readability

File: pkg/upstream/cluster/edfheap.go [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+		}
+	}
+	if oldI == i {
+		return false
+	}
+	h.elements[i] = element
+	return true
+}
+
+func (h *edfHeap) fixDown(i, n int) bool {
```
Comment (alpha-baby) [inline pkg/upstream/cluster/edfheap.go:104]: need code readability

File: pkg/fastheap/fastheap_test.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
+/*
```
Comment (alpha-baby) [inline pkg/fastheap/fastheap_test.go:1]: delete not use

File: pkg/fastheap/fastheap.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
+/*
```
Comment (alpha-baby) [inline pkg/fastheap/fastheap.go:1]: delete not use

File: pkg/upstream/cluster/edfheap_test.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (alpha-baby) [inline pkg/upstream/cluster/edfheap_test.go:1]: 把官方包里面的单元测试和你写的这个对齐一下吧

File: pkg/upstream/cluster/edfheap.go (hunk 2 above)
Comment (alpha-baby) [reply pkg/upstream/cluster/edfheap.go:142]: should:

```go
if minCap < h.elements && h.size*2 < cap(h.elements) {
//.....
}
```

**Review by alpha-baby — DISMISSED** [review]
LGTM

File: pkg/upstream/cluster/edfheap.go [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+func (h *edfHeap) Fix(i int) {
+	if !h.fixDown(i, h.size) {
+		h.fixUp(i)
+	}
+}
+
+func (h *edfHeap) Push(element *edfEntry) {
+	// TODO: It seems don't need to change the slice capacity dynamically?
+	h.ensureIncrement()
```
[the remark this answers is not in the document]
Comment (alpha-baby) [reply pkg/upstream/cluster/edfheap.go:54]: There is indeed no Expansion and reduction  of slice capacity in the edf algorithm or in the current scenario within mosn. @jizhuozhi

[the remark this answers is not in the document]
Comment (jizhuozhi) [reply pkg/upstream/cluster/edfheap.go:54]: No problem, I will submit the revised code later @doujiang24 @alpha-baby

**Comment by jizhuozhi** [comment]
Hi, I have submitted the latest code, please review it @alpha-baby @doujiang24

**Review by alpha-baby — APPROVED** [review]
good job!

**Review by doujiang24 — APPROVED** [review]

**Pull request merged**